### DISCUSSION TITLE
How to send data from zwavejs2mqtt inside a docker container to a Mosquitto server?

### DISCUSSION THREAD

**Discussion by ZeroGravityByUs** [zwave-js/zwave-js-ui#1244] [Q&A] [answered]
I'm now running **zwavejs2mqtt** on a Raspberry Pi, in a docker container, and it's working just fine!
When ever I push my Z-Wave button, I can see the response in the terminal!! Hurray! 

```
2021-05-18 17:38:55.733 INFO ZWAVE: Node 2: value updated: 113-0-Home Security-Sensor status 2 => 2  
2021-05-18 17:38:55.760 INFO ZWAVE: Node 2: value notification: 91-0-scene-001 0  
```       

But how can I get this data in and out of the container, to a Mosquitto server for example?

I have the Mosquitto server running on my Raspberry Pi on port 1883, and if I try to open up the port 1883 in the docker container using **-p 1883:1883**, I receive an error that the port is already in use? What am I'm doing wrong?

```
docker run --rm -it -p 8091:8091 -p 3000:3000 -p 1883:1883 --device=/dev/ttyACM0 -v $(pwd)/store:/usr/src/app/store zwavejs/zwavejs2mqtt:latest
```  

Error: 
```
docker: Error response from daemon: driver failed programming external connectivity on endpoint quirky_banzai (cb39ca10887ef1290fe659391a24e79740a26e6f7eb6df4f443c84e103238cf0): Error starting userland proxy: listen tcp4 0.0.0.0:1883: bind: address already in use.
```

PS. I'm using a Raspberry Pi 3 Model B Rev 1.2 and my goal is (to put it simple) to show a webpage when my Z-Wave button is pressed once, and then show another webpage when my Z-Wave button is pressed once more!

**Comment by robertsLando** [comment, marked as answer]
Hummm that's not the way to do it. With `-p 1883:1883` you are mapping z2m container port to your local port 1883 but that is actually used by mosquitto. You have two alternatives, the first one is to map the container network to your local network the second one is to use a container also for mosquitto so you can setup the communication between z2m and mosquitto containers (using a docker compose file).

This example uses a docker compose file to map containers network to local net: https://tewarid.github.io/2019/05/24/docker-compose-to-run-node-red-and-mosquitto.html#:~:text=Mosquitto%20can%20be%20accessed%20at%20port%201883%20on%20the%20host%20network.

**Reply by ZeroGravityByUs** [reply]
Hi Robert,  
Thanks for your quick answer! I'm very new to Docker as you might have guessed by now... 
[I got some help regarding how to install zwavejs2mqtt using snap](
https://github.com/zwave-js/zwavejs2mqtt/discussions/1216), so I will continue along that route for now.   
But I'm very thankful for your answer.
